File: Leetcode_Python/numIslands_bfs.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        islands = 0
        rows = len(grid)
        cols = len(grid[0])

        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]

        def bfs(row, col):
            queue = deque()
            grid[row][col] = '0'
            queue.append((row,col))

            while queue:
                row, col = queue.popleft()
                for r, c in directions:
                    newRow = row + r
                    newCol = col + c

                    if(newRow < 0 or newRow >= rows or newCol < 0 or newCol >= cols or grid[newRow][newCol] == '0'):
                        continue
                    
                    queue.append((newRow, newCol))
                    grid[newRow][newCol] = '0'

        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == '1':
                    bfs(row, col)
                    islands += 1

        return islands
        

File: Leetcode_Java/numIslands_bfs.java
import java.util.*;

class Solution {
    public int numIslands(char[][] grid) {
        int islands = 0;
        int rows = grid.length;
        int cols = grid[0].length;

        int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                if (grid[row][col] == '1') {
                    bfs(grid, row, col, directions);
                    islands++;
                }
            }
        }

        return islands;
    }

    private void bfs(char[][] grid, int row, int col, int[][] directions) {
        int rows = grid.length;
        int cols = grid[0].length;

        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{row, col});
        grid[row][col] = '0'; // mark as visited

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int currRow = current[0];
            int currCol = current[1];

            for (int[] dir : directions) {
                int newRow = currRow + dir[0];
                int newCol = currCol + dir[1];

                if (newRow >= 0 && newRow < rows &&
                    newCol >= 0 && newCol < cols &&
                    grid[newRow][newCol] == '1') {

                    queue.offer(new int[]{newRow, newCol});
                    grid[newRow][newCol] = '0'; // mark as visited
                }
            }
        }
    }
}